### PR TITLE
NI-DKG V0

### PR DESCRIPTION
**Pull request by benr-ml**

### CODE REVIEW

File: fastcrypto-tbls/src/nidkg.rs [hunk 1]
```diff
@@ -0,0 +1,493 @@
...
+                    .expect("using valid share id");
+                let encryptions = (0..NUM_OF_ENCRYPTIONS_PER_SHARE)
+                    .into_iter()
+                    .map(|_| {
+                        let r = G::ScalarType::rand(rng);
+                        let msg = bcs::to_bytes(&r).expect("serialization should work");
+                        let r_g = G::generator() * r;
+                        let r_x_g = *node.pk.as_element() * r;
+                        // Save also the points instead of recomputing them later.
+                        values.push((r, r_x_g));
```
Comment (jonas-lj) [inline fastcrypto-tbls/src/nidkg.rs:177]: I think the values vector is not needed - the values can be pushed to interim_values directly.

File: fastcrypto-tbls/src/dl_verification.rs [hunk 2]
```diff
@@ -1,13 +1,126 @@
...
+        .map(|(i, c)| *c * poly_f.eval(NonZeroU32::new((i + 1) as u32).unwrap()).value)
+        .collect::<Vec<_>>();
+    let lhs = G::multi_scalar_mul(&coefficients[..], values).expect("sizes match");
+    if lhs != G::zero() {
+        return Err(FastCryptoError::InvalidProof);
+    }
+    Ok(())
+}
+
+/// Checks if vectors v1=(a1*G1, ..., an*G1) and v2=(a1'*G2, ..., an'*G2) use ai = ai' for all i, by
```
[the remark this answers is not in the document]
Comment (benr-ml) [reply fastcrypto-tbls/src/dl_verification.rs:102]: the inputs are indeed vectors, and they can be represented using multiplications with one base element. the vectors are v1 and v2 (renamed the arguments to the function)

File: fastcrypto-tbls/src/nidkg.rs (hunk 1 above)
Comment (benr-ml) [reply fastcrypto-tbls/src/nidkg.rs:177]: note that interim_values is a vector of fixed size arrays

File: fastcrypto-tbls/src/dl_verification.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jonas-lj) [reply fastcrypto-tbls/src/dl_verification.rs:104]: Got it. The naming confused me.

File: fastcrypto/src/groups/bls12381.rs [hunk 3]
```diff
@@ -649,6 +650,22 @@ impl ScalarType for Scalar {
...
+    fn hash_to_group_element(msg: &[u8]) -> Self {
+        let hashed = Sha512::digest(msg).digest;
+        let mut ret = blst_fr::default();
+        let mut bytes = [0u8; SCALAR_LENGTH];
+        bytes.copy_from_slice(&hashed[..SCALAR_LENGTH]);
+        // TODO: is there an overflow?
+        unsafe {
+            let mut scalar = blst_scalar::default();
+            blst_scalar_from_bendian(&mut scalar, bytes.as_ptr());
+            blst_fr_from_scalar(&mut ret, &scalar);
```
Comment (jonas-lj) [inline fastcrypto/src/groups/bls12381.rs:663]: Since the modular reduction is done from a blst_scalar (which is 256 bits), the result will be biased.

Comment (benr-ml) [reply fastcrypto/src/groups/bls12381.rs:664]: Replaced with FiatShamirChallenge

File: fastcrypto-tbls/src/dl_verification.rs [hunk 4]
```diff
@@ -60,6 +61,120 @@ pub fn verify_poly_evals<G: GroupElement + MultiScalarMul, R: AllowedRng>(
...
+) -> FastCryptoResult<()> {
+    if v1.len() != v2.len() {
+        return Err(FastCryptoError::InvalidProof);
+    }
+    let rs = get_random_scalars::<bls12381::Scalar, R>(v1.len() as u32, rng);
+    let lhs = bls12381::G1Element::multi_scalar_mul(&rs[..], v1).expect("sizes match");
+    let rhs = bls12381::G2Element::multi_scalar_mul(&rs[..], v2).expect("sizes match");
+
+    if lhs.pairing(&bls12381::G2Element::generator())
+        != bls12381::G1Element::generator().pairing(&rhs)
```
Comment (jonas-lj) [inline fastcrypto-tbls/src/dl_verification.rs:171]: I don't know if it's critical here, but pairings with constant input from G_2 may be done faster. In blst it's done via blst's `precompute_lines` function.

**Review by jonas-lj — APPROVED** [review]

File: fastcrypto-tbls/src/dl_verification.rs (hunk 4 above)
Comment (benr-ml) [reply fastcrypto-tbls/src/dl_verification.rs:171]: it's not a dominant operation in this case, so leaving it as is

**Pull request merged**